flink-runtime/src/test/java/org/apache/flink/runtime/testutils/CommonTestUtils.java

```diff
@@ -182,8 +182,10 @@ public static void waitForAllTaskRunning(
                                                             .allMatch(
                                                                     task ->
                                                                             task.getExecutionState()
-                                                                                    == ExecutionState
-                                                                                            .RUNNING));
+                                                                                            == ExecutionState
+                                                                                                    .RUNNING
+                                                                                    || task.getExecutionState()
+                                                                                            .isTerminal()));
```

Comment:
I would create a separate JIRA ticket for this (if it brakes something, it would be easier to find/analyse/reference it). It sounds like it's indeed a valid fix.

However shouldn't we throw an exception if state is terminal? `waitForAllTaskRunning` suggests that this is an unexpected outcome.

Reply:
That's a good point.

I think we should NOT throw an exception because if the task is in a terminal state it can later become running again after a failover for example.

This is possible because there are two checks:
- graph is running &&
- task is running

which are evaluated in a loop.

I'll create a separate ticket.

Reply:
I've created FLINK-23811 / #16845 to address this (will base this PR on that change).